### ISSUE TITLE
CPTGraph initWithCoder: doesn't properly remove/add plot spaces

### ISSUE THREAD

**Issue by boisy** [core-plot/core-plot#363]
I'm using CPTGraph in an application that saves/loads NSDocuments. In saving/loading the graph via the NSCoding protocol methods, I noticed that a loaded graph would no longer respond to changes in the plotspace, whereas a newly created document with a new graph would.

I inspected the code path of the initWithCoder: method in CPTGraph and discovered that the call stack eventually goes into CPTGraph's initWithFrame: method. There, a new plot space is created and the CPTGraph's addPlotSpace: method is called. That method setups up CPTGraph as an observer of changes to the plot space mapping.

All well and good until we get back into CPTGraph's initWithCoder: method, where the following code is executed:

```
        CPTPlotSpaceArray *plotSpaceArray = [coder decodeObjectOfClasses:[NSSet setWithArray:@[[NSArray class], [CPTPlotSpace class]]]
                                                                  forKey:@"CPTGraph.plotSpaces"];
        if ( plotSpaceArray ) {
            plotSpaces = [plotSpaceArray mutableCopy];
        }
        else {
            plotSpaces = [[NSMutableArray alloc] init];
        }
```

The line with the mutableCopy sets the CPTGraph's plotSpaces property to the decoded array above. 

This assignment essentially wipes out the earlier plot space which was observed. The assignment also fails to register CPTGraph as an observer of this new plot space.

As I test, I altered the code as follows, and am getting proper responses when decoding a CPTGraph:

```
        CPTPlotSpaceArray *plotSpaceArray = [coder decodeObjectOfClasses:[NSSet setWithArray:@[[NSArray class], [CPTPlotSpace class]]]
                                                                  forKey:@"CPTGraph.plotSpaces"];
        if ( plotSpaceArray ) {
            for (CPTPlotSpace *space in self.plotSpaces) {
                [self removePlotSpace:space];
            }
            
            for (CPTPlotSpace *space in plotSpaceArray) {
                [self addPlotSpace:space];
            }
        }
        else {
            plotSpaces = [[NSMutableArray alloc] init];
        }
```

In this code, the old plot spaces are removed properly, allowing them to be unobserved in the removePlotSpace method. Additionally, the new plot spaces are added via addPlotSpace: which properly registers them to be observed.

I'm not certain if this is the best approach, but it does address the issue I was seeing.

**Comment by eskroch** [comment]
Fixed on the `release-2.3` branch.

I modified your solution slightly. Since this is an `init` method, we always need to initialize the plot space array and there's no need to remove the old spaces since the array will always be empty before we add the new plot spaces from the coder.

**Comment by boisy** [comment]
Awesome! Thanks!
